scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala

```diff
@@ -296,6 +296,65 @@ object TreeSyntax {
       }
       !isOnlyChildOfOnlyChild(t)
     }
+    def guessNeedsLineSep(t: Template): Boolean =
+      t.stats.isEmpty
+    @annotation.tailrec
+    def guessNeedsLineSep(t: Stat): Boolean = t match {
+      case _: Decl.Val => true
+      case _: Decl.Var => true
+      case _: Decl.Def => true
+      case _: Decl.Type => false
+      case Defn.Val(_, _, _, rhs) => guessNeedsLineSep(rhs)
+      case Defn.Var.After_4_7_2(_, _, _, body) => guessNeedsLineSep(body)
+      case Defn.Def.After_4_7_3(_, _, _, _, body) => guessNeedsLineSep(body)
+      case _: Defn.Macro => true
+      case _: Defn.Type => true
+      case Defn.Class.After_4_6_0(_, _, _, _, template) => guessNeedsLineSep(template)
+      case Defn.Trait.After_4_6_0(_, _, _, _, template) => guessNeedsLineSep(template)
+      case Defn.Object(_, _, template) => guessNeedsLineSep(template)
+      case _: Term.This => true
```

Comment:
I checked every single case by hand on a Scala 2.13.10 test project. Unless I'm missing something `this` with a block after does not compile:
```scala
object Main9 {
  def main(args: Array[String]): Unit = {
    this
    {
      val a: Int = 0
      a
    }
  }
}
```
```
sbt:test> compile
[info] compiling 10 Scala sources and 1 Java source to C:\Scala\test\target\scala-2.13\classes ...
[error] C:\Scala\test\src\main\scala\Main9.scala:4:5: Main9.type does not take parameters
[error]     {
[error]     ^
[error] one error found
[error] (Compile / compileIncremental) Compilation failed
```

Reply:
is this interpreted as an invocation of a secondary ctor? :)

Reply:
I think it's something like that, but q"this; {a}" requires a new line (or the ';').
Because even if there is a way for `this\n{a}` to compile, it's still not the original expression for sure :)